src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,98 @@
+package duke;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+
+
+/**
+ * Handles file state.
+ */
+public class Storage {
+    private static Map<String, Storage> fileStates = new HashMap<>();
+    private String contents = null;
+    private String fileName;
+
+    private Storage(String fileName) {
+        this.fileName = fileName;
+    }
+
+    /**
+     * Gets a singleton object that manages a particular file.
+     *
+     * @param fileName File name to identify file.
+     * @return Object managing the file.
+     */
+    public static Storage getFileState(String fileName) {
+        if (!fileStates.containsKey(fileName)) {
+            fileStates.put(fileName, new Storage(fileName));
+        }
+        return fileStates.get(fileName);
+    }
+
+    private static String convertToBase64(String input) {
+        return Base64.getEncoder().encodeToString(input.getBytes(StandardCharsets.UTF_16));
+    }
+
+    private static String convertFromBase64(String input) {
+        return StandardCharsets.UTF_16.decode(ByteBuffer.wrap(Base64.getDecoder().decode(input))).toString();
+    }
+
+    /**
+     * Gets lines from file as a String[][].
+     *
+     * @return Array of String[], each storing comma-separated parts of a line.
+     */
+    public String[][] getLines() {
+        ArrayList<String[]> lines = new ArrayList<>();
+        try {
+            File f = new File(fileName);
+            Scanner sc = new Scanner(f);
+            while (sc.hasNext()) {
+                ArrayList<String> curLine = new ArrayList<>();
+                for (String str : sc.nextLine().split(",")) {
+                    curLine.add(convertFromBase64(str));
+                }
+                lines.add(curLine.toArray(new String[]{}));
+            }
+            return (String[][]) lines.toArray(new String[][]{});
+        } catch (FileNotFoundException ex) {

```

Comment:
Everything is good!!! Just one small suggestion that maybe you can set up some exceptions classes, (e.g. DukeException extends Exception) and put all exception class into one package : )

Reply:
The existing exceptions I find a pretty useful, so no need to duplicate them. Perhaps a Response class to ensure that we print only once can be useful. Perhaps having it throwable like an exception can help too, so that once it is decided it can skip past the happy path to land in a catch block.

Reply:
Done as stated. It was quite a hassle.